sdks/python/apache_beam/runners/dataflow/dataflow_runner.py

```diff
@@ -906,14 +924,40 @@ def wait_until_finish(self, duration=None):
       thread.start()
       while thread.isAlive():
         time.sleep(5.0)
-      if self.state != PipelineState.DONE:
-        # TODO(BEAM-1290): Consider converting this to an error log based on the
-        # resolution of the issue.
+
+      terminated = self._is_in_terminal_state()
+      assert duration or terminated, (
+          'Job did not reach to a terminal state after waiting indefinitely.')
+
+      if terminated and self.state != PipelineState.DONE:
+        # TODO(BEAM-1290): Consider converting this to an error log based on
+        # theresolution of the issue.
         raise DataflowRuntimeException(
             'Dataflow pipeline failed. State: %s, Error:\n%s' %
             (self.state, getattr(self._runner, 'last_error_msg', None)), self)
     return self.state
 
+  def cancel(self):
+    if not self.has_job:
+      raise IOError('Failed to get the Dataflow job id.')
+
+    self._update_job()
+
+    if self._is_in_terminal_state():
+      logging.warning(
+          'Cancel failed because job %s is already terminated in state %s.',
+          self.job_id, self.state)
+    else:
+      if not self._runner.dataflow_client.modify_job_state(
+          self.job_id, 'JOB_STATE_CANCELLED'):
```

Comment:
<!--thread_id:cc_135128285_t; commit:5b7323ad9e10a5693ea41a239c8a675139b977de; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> This should be `self.job_id()` since the method is not a `@property`.

<!--section:body-->
Done.